### ISSUE TITLE
SC2260 and SC3021 false positives

### ISSUE THREAD

**Issue by alkisg** [koalaman/shellcheck#2514]
I believe the following code is correct, yet it triggers SC2260 and SC3021:

```bash
#!/bin/sh

for fd in 1 2; do
    echo Hello >&"$fd" | grep -o H
done
```

The shellcheck output is:

```console
$ shellcheck /tmp/test.sh

In /tmp/test.sh line 4:
    echo Hello >&"$fd" | grep -o H
               ^-- SC2260 (error): This redirection overrides the output pipe. Use 'tee' to output to both.
               ^-----^ SC3021 (warning): In POSIX sh, >& is undefined.

For more information:
  https://www.shellcheck.net/wiki/SC2260 -- This redirection overrides the ou...
  https://www.shellcheck.net/wiki/SC3021 -- In POSIX sh, >& is undefined.
```

**Comment by elsandosgrande** [comment]
Your shebang says `sh`, not `bash`, hence the Shellcheck messages. Shellcheck treats `sh` as the original POSIX shell for compatibility purposes, so you need to explicitly state that your script is a BASH script and not a fully POSIX-compliant shell script.

**Comment by alkisg** [comment]
My script runs fine with dash and posh, is what I wrote non-POSIX compliant?
I'm not trying to write something bash-specific...

**Comment by elsandosgrande** [comment]
It's not that file descriptor redirection isn't a thing in other shells, it just doesn't seem to be a thing in the original POSIX shell, hence why it says that `>&` is undefined in POSIX. At any rate, running it through BASH and ZSH gives me different results, with the BASH result seeming the most accurate to what I'd expect:

BASH and `sh` shebangs
```
$ ./script.sh
H
Hello
```

ZSH shebang
```
$ ./script.sh
H
H
Hello
H
```

At any rate, it seems that `grep` only picks up on the first file descriptor, which could be related to the error above. Also, when I set the shebang to DASH, Shellcheck says that DASH doesn't support file descriptor redirection outright:

![image](https://user-images.githubusercontent.com/40576902/173255480-24696226-454d-4384-ba6e-9e663553cece.png)

**Comment by alkisg** [comment]
Posix defines output file redirection [there](https://pubs.opengroup.org/onlinepubs/9699919799/utilities/V3_chap02.html#tag_18_07_06):

> The redirection operator:
> [n]>&word
> shall duplicate one output file descriptor from another, or shall close one. **If word evaluates to one or more digits**, the file descriptor denoted by n,

I think the part that I made bold means that `word` can be a variable, like in my script, where $fd is 1 or 2.

Shellcheck doesn't complain about redirection in dash when constants are used:

```console
$ cat script.sh
#!/bin/dash
echo Hello >&1 | grep -o H
echo Hello >&2 | grep -o H

$ shellcheck script.sh 
<no output>
```

It only complains when `word` is a variable. I think by mistake it's expecting "a number" there instead of "an expression", and when it doesn't see a number, it thinks it's a different thing, a bashism, and that's why it says it's not supported in dash.

Zsh defines redirection [there](https://zsh.sourceforge.io/Doc/Release/Redirection.html).
I'm not sure if the correct behavior do is to move the fd or dup the fd, but I think it's not a syntax error in any of the shells...

**Comment by elsandosgrande** [comment]
Well then, I stand corrected 😅. Either way though, the Shellcheck error says that redirection and piping can interfere with each other and thus to stick to one or the other in a given string of commands. That said, I'm not sure if that second `Hello` is due to that issue or due to `grep` not reading from standard error 😅

**Comment by alkisg** [comment]
Sure, the minimal example doesn't do anything useful, of course that's not what I was trying to do in my real scripts. :rofl:
But if you replace the variable digit with a constant digit, shellcheck doesn't complain anymore; it's the "variable syntax" that it doesn't like, not the semantics/goal/usefullness of the script...

**Comment by bhaible** [comment]
See also issue https://github.com/koalaman/shellcheck/issues/2520.

**Comment by koalaman** [comment]
You no longer get a SC3021 thanks to #2520, but you do still get a SC2260 about overriding the pipe which still seems correct. Ideally it would also warn about `>&- | cmd` but ShellCheck is all about heuristics and these are not the kinds of things people do unintentionally.
